src/amber/controller/helpers/responders.cr

```diff
@@ -31,40 +41,46 @@ module Amber::Controller::Helpers
       end
 
       def type
-        select_type.to_s
+        (@type ||= select_type).to_s
       end
 
       def body
-        @available_responses[type]
+        @available_responses[type]?
       end
 
       private def select_type
-        @type ||= begin
-          raise "You must define at least one response_type." if @available_responses.empty?
-          @requested_responses << @available_responses.keys.first
-          @requested_responses.find do |resp|
-            @available_responses.keys.includes?(resp)
-          end
+        raise "You must define at least one response_type." if @available_responses.empty?
+        # NOTE: If only one response is requested don't return something else.
+        @requested_responses << @available_responses.keys.first if @requested_responses.size != 1
+        @requested_responses.find do |resp|
+          @available_responses.keys.includes?(resp)
         end
       end
     end
 
-    private def requested_responses
-      req_responses = Array(String).new
+    private def extension_request_type
+      path_ext = request.path.match(Content::TYPE_EXT_REGEX).try(&.[1])
+      return [Content::TYPE[path_ext]] if path_ext
+    end
 
-      if (accept = context.request.headers["Accept"]?) && !accept.empty?
-        accepts = accept.split(";").first?.try(&.split(/,|,\s/))
-        req_responses.concat(accepts) if accepts.is_a?(Array) && accepts.any?
+    private def accepts_request_type
+      accept = context.request.headers["Accept"]?
+      if accept && !accept.empty?
+        accepts = accept.split(";").first?.try(&.split(Content::ACCEPT_SEPARATOR_REGEX))
+        return accepts if !accepts.nil? && accepts.any?
       end
-      req_responses
+    end
+
+    private def requested_responses
+      extension_request_type || accepts_request_type || [] of String
     end
 
     protected def respond_with(&block)
       content = with Content.new(requested_responses) yield
       if content.body
         set_response(body: content.body, status_code: 200, content_type: content.type)
       else
-        set_response(body: "Response unexceptable.", status_code: 406, content_type: Content::TYPE[:text])
+        set_response(body: "Response Not Acceptable.", status_code: 406, content_type: Content::TYPE[:text])
```

Comment:
Hi @elorest I didn't see it 😅 

Why not `raise Amber::Exception::NoAcceptable` so we can use `Amber::Pipe::Error` here 👉 https://github.com/amberframework/amber/blob/master/src/amber/router/pipe/error.cr#L19 and be hable to customize the 406 reponse. WDYT?

Reply:
I was under the impression that we didn't want to raise exceptions because it was slower or something. I remember @eliasjpr saying that was the reason he made halt instead of just raising and exception to handle later. Also since I'm setting a response anyone it sort of seems natural to set the other response and message.